pkg/frontend/v2/remoteexec.go

```diff
@@ -385,3 +393,154 @@ func accountForHPointMemoryConsumption(d []promql.HPoint, memoryConsumptionTrack
 
 	return nil
 }
+
+type eagerLoadingResponseStream struct {
+	inner       responseStream
+	bufferedErr error
+
+	mtx              *sync.Mutex
+	buffer           *responseStreamBuffer
+	newDataAvailable chan struct{}
+}
+
+func newEagerLoadingResponseStream(ctx context.Context, inner responseStream) *eagerLoadingResponseStream {
+	stream := &eagerLoadingResponseStream{
+		inner:  inner,
+		mtx:    &sync.Mutex{},
+		buffer: &responseStreamBuffer{},
+	}
+
+	go stream.startBuffering(ctx)
+
+	return stream
+}
+
+func (e *eagerLoadingResponseStream) startBuffering(ctx context.Context) {
+	for {
+		if !e.bufferOne(ctx) {
+			return
+		}
+	}
+}
+
+func (e *eagerLoadingResponseStream) bufferOne(ctx context.Context) bool {
+	msg, err := e.inner.Next(ctx)
+
+	e.mtx.Lock()
+	defer e.mtx.Unlock()
+	defer e.notifyNewDataAvailable()
+
+	e.buffer.Push(bufferedMessage{msg, err})
+
+	return err == nil
+}
+
+// notifyNewDataAvailable unblocks a pending Next() call.
+// It must only be called with e.mtx locked.
+func (e *eagerLoadingResponseStream) notifyNewDataAvailable() {
+	if e.newDataAvailable == nil {
+		return
+	}
+
+	close(e.newDataAvailable)
+	e.newDataAvailable = nil
+}
+
+func (e *eagerLoadingResponseStream) Next(ctx context.Context) (*frontendv2pb.QueryResultStreamRequest, error) {
+	if e.bufferedErr != nil {
+		return nil, e.bufferedErr
+	}
+
+	e.mtx.Lock()
+
+	if e.newDataAvailable != nil {
+		e.mtx.Unlock()
+		return nil, errMultipleEagerLoadingNextCalls
+	}
+
+	if e.buffer.Any() {
+		msg := e.buffer.Pop()
+		defer e.mtx.Unlock()
+
+		if msg.err != nil {
+			e.bufferedErr = msg.err
+		}
+
+		return msg.payload, msg.err
+	}
+
+	newDataAvailable := make(chan struct{}) // We can't store this directly in e.newDataAvailable because e.newDataAvailable might be cleared by the time we wait on it below.
```

Comment:
(non-blocking) Don't really understand this part, just making a note so I can ask about it next time.

Reply:
Could you elaborate? Are you referring to why we don't store this in `e.newDataAvailable`?